Activities/Database/UiPath.Database/DatabaseConnection.cs

```diff
@@ -94,6 +99,182 @@ public virtual int InsertDataTable(string tableName, DataTable dataTable)
+
+                // Perform an initial count on the destination table.
+                DbCommand commandRowCount = _connection.CreateCommand();
+                commandRowCount.Connection = _connection;
+                commandRowCount.CommandText = string.Format("SELECT COUNT(*) FROM {0}", tableName);
+
+                long countStart = Convert.ToInt32(
+                    commandRowCount.ExecuteScalar());
+
+                List<string> gacFolders = new List<string>() {
+                    "GAC", "GAC_32", "GAC_64", "GAC_MSIL",
+                    "NativeImages_v2.0.50727_32",
+                    "NativeImages_v2.0.50727_64",
+                    "NativeImages_v4.0.30319_32",
+                    "NativeImages_v4.0.30319_64"
+                };
+                List<string> oracleList = new List<string>();
+                foreach (string folder in gacFolders)
+                {
+                    string path = Path.Combine(
+                       Environment.ExpandEnvironmentVariables(@"%systemroot%\assembly"),
+                       folder);
+
+                    if (Directory.Exists(path))
+                    {
+
+                        string[] assemblyFolders = Directory.GetDirectories(path);
+                        foreach (string assemblyFolder in assemblyFolders)
+                        {
+                            string[] files = Directory.GetFiles(assemblyFolder, "oracle.manageddataaccess*", SearchOption.AllDirectories);
+                            oracleList.AddRange(files);
+                        }
+                    }
+
+                    path = Path.Combine(
+                       Environment.ExpandEnvironmentVariables(@"%systemroot%\Microsoft.NET\assembly"),
+                       folder);
+
+                    if (Directory.Exists(path))
+                    {
+
+                        string[] assemblyFolders = Directory.GetDirectories(path);
+                        foreach (string assemblyFolder in assemblyFolders)
+                        {
+                            string[] files = Directory.GetFiles(assemblyFolder, "oracle.manageddataaccess*", SearchOption.AllDirectories);
+                            oracleList.AddRange(files);
+                        }
+                    }
+                }
+
+                if (oracleList.Count > 0)
+                {
+                    bool bulkCopyIsPresent = false;
+                    //try all drivers with latest version first
+                    foreach (var item in oracleList.OrderByDescending(x => x))
+                    {
+
+                        var oracle = Assembly.LoadFile(item);
+                        var bulkCopyType = oracle.GetType("Oracle.ManagedDataAccess.Client.OracleBulkCopy");
+
+                        if (bulkCopyType != null)
+                        {
+                            dynamic bulkCopy = Activator.CreateInstance(bulkCopyType, new object[] { connection });
```

Comment:
same here. if make some utility interfaces we could test most of the logic. all we need to abstract is the call that goes to the db (WriteToServer in this case)

Reply:
extracted to interface